pkg/executor/sortexec/topn.go

```diff
@@ -166,76 +211,312 @@ func (e *TopNExec) loadChunksUntilTotalLimit(ctx context.Context) error {
 			break
 		}
 		e.chkHeap.rowChunks.Add(srcChk)
+		if e.spillHelper.isSpillNeeded() {
+			e.isSpillTriggeredInStage1ForTest = true
+			break
+		}
+
+		injectTopNRandomFail(1)
 	}
-	e.initPointers()
-	e.initCompareFuncs()
-	e.buildKeyColumns()
+
+	e.chkHeap.initPtrs()
 	return nil
 }
 
 const topNCompactionFactor = 4
 
-func (e *TopNExec) executeTopN(ctx context.Context) error {
-	heap.Init(e.chkHeap)
-	for uint64(len(e.chkHeap.rowPtrs)) > e.totalLimit {
-		// The number of rows we loaded may exceeds total limit, remove greatest rows by Pop.
-		heap.Pop(e.chkHeap)
-	}
+func (e *TopNExec) executeTopNNoSpill(ctx context.Context) error {
 	childRowChk := exec.TryNewCacheChunk(e.Children(0))
 	for {
+		if e.spillHelper.isSpillNeeded() {
+			e.isSpillTriggeredInStage2ForTest = true
+			return nil
+		}
+
 		err := exec.Next(ctx, e.Children(0), childRowChk)
 		if err != nil {
 			return err
 		}
+
 		if childRowChk.NumRows() == 0 {
 			break
 		}
-		err = e.processChildChk(childRowChk)
-		if err != nil {
-			return err
-		}
+
+		e.chkHeap.processChkNoSpill(childRowChk)
+
 		if e.chkHeap.rowChunks.Len() > len(e.chkHeap.rowPtrs)*topNCompactionFactor {
-			err = e.doCompaction(e.chkHeap)
+			err = e.chkHeap.doCompaction(e)
 			if err != nil {
 				return err
 			}
 		}
+		injectTopNRandomFail(10)
+	}
+
+	slices.SortFunc(e.chkHeap.rowPtrs, e.chkHeap.keyColumnsCompare)
+	return nil
+}
+
+func (e *TopNExec) spillRemainingRowsWhenNeeded() error {
+	if e.spillHelper.isSpillTriggered() {
+		return e.spillHelper.spill()
+	}
+	return nil
+}
+
+func (e *TopNExec) checkSpillAndExecute() error {
+	if e.spillHelper.isSpillNeeded() {
+		// Wait for the stop of all workers
+		e.fetcherAndWorkerSyncer.Wait()
```

Comment:
So spilling disk is in single concurrent?
Why not let the workers hold separate heaps and do sort and spill separately?

Reply:
And then there is no need for `fetcherAndWorkerSyncer`